Add MinimalSurface function

vladimir-ch opened the pull request:
The first truly interesting function to be added to functions subpackage. Taken from MINPACK-2 test problem collection, good for testing our medium/large-scale minimization capabilities (although the function itself does not seem to be difficult to minimize). We can also use it to decide at which problem dimension we should switch the default method from BFGS to LBFGS. And it is also a good example of how to implement non-trivial functions.

vladimir-ch commented:
Finished polishing the code. PTAL, @btracey

btracey commented on functions/minsurf.go:
```diff
@@ -0,0 +1,245 @@
+	return ms.hx, ms.hy
+}
+
+// WriteGnuplot saves in Gnuplot format the surface represented by the vector
```
If we keep the function in this format, it should take in a reader, and do error checking on the writes. It seems better instead to return the surface as an XYZer of some kind (like in plot). This would enable use with Plot and we could add a function to Plot to produce a gnuplot file from an XYZer. I would suggest something like
Surface(x []float64) []XYZ {}

where XYZ is a struct{X, Y, Z float64) and implements the plot.XYZer interface

kortschak replied:
Rather, can I suggest you fill a mat64.Dense and two slices, one for row coordinates and one for column coordinates. These can all be put into a struct that is an XYZer, is more space efficient and will work with my heatmap and contour code. Free form XYZer representation won't.

kortschak replied:
Even better, make a type  that dynamically returns the X or Y values. Basically a float64 derived type that does func (v vector) At(i int) float64 { return -0.5 + float64(i)*float64(v) } where type vector float64.

I'm still fleshing out how I want this interface to look for the surface plotting, but if you do these you get XYZer. It needs to be more subtle because, at least with my code, an XYZer is not adequate since the X and Y values must be on a grid, which XYZer doesn't guarantee.

Maybe wrap it all up in one type that is an XYZer and add to more methods, Matrix() mat64.Matrix and Coordinates() (x, y Vector) (name to be determined, but which implements the At method above and a Len() int method - although this is strictly unnecessary since Dims, but nice to be complete).

btracey replied:
Maybe we should just remove this until we know how it fits into the bigger plotting picture.

kortschak replied:
I have a good interface.

```
type MeshGrid interface {
    Dims() (r, c int)
    At(r, c) float64
    XCoord(int) float64
    YCoord(int) float64
}
```

This can be wrapped up to make an XYZer if needed and would be implementable with the types I described above.

vladimir-ch replied:
* MeshGrid seems to be rather a function defined on a grid, wouldn't GridFunction be a better name?
* It would be more natural for Dims and At to use x, y rather than y, x.

I will remove WriteGnuplot().

kortschak replied:
I'm going to push my WIP changes for plotting surfaces soon, you can see what I mean there.

I see your point, but it complicates the common case where a grid is defined by a matrix. Currently, that is a three line shim between mat64.Matrix and GridMesh.

vladimir-ch replied:
Ah, ok, I see what you mean.

kortschak replied:
When I have finished up the contour plotter I'll make a sub branch with the GridFunction approach so they can be compared. I'm thinking for the GridFunction approach that it would have `Z(c, r int) float64`, `X(c int) float64` and `Y(r int) float64` along with `Dims() (c, r int)`.

btracey commented on functions/minsurf_test.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/gonum/floats"
+)
+
+func TestMinimalSurface(t *testing.T) {
```
Why doesn't this use testFunction?

vladimir-ch replied:
For two reasons
1. The only alternative implementation I have is the 20+ years old MINPACK code. I did not feel like fiddling with it to make it compile, run and extract a reference solution.
2. Even if I had a reference X, I did not want to dump those large vectors into the code. I admit that this is not a serious reason.

btracey replied:
Okay, this is fine. If we add more functions where this is the case (and I imagine we will), we should break this off into a testGradient function.

vladimir-ch replied:
Strictly speaking, I shouldn't be testing if the gradient at ExactX() is small (especially at such coarse grid resolutions), because ExactX is the exact solution projected on the grid, not the exact solution to the discrete problem. I will reconsider it after the weekend.

vladimir-ch commented:
* Removed `WriteGnuplot()` method
* Added a comment to the test which checks that the gradient at `ExactX()` is small
* I would add a `testGradient()` function later when the usage pattern is clearer

PTAL, @btracey

vladimir-ch commented:
When `plot.GridFunc` interface is settled and merged, I can add a method that returns something that satisfies it.

btracey commented:
LGTM
